cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll

```diff
@@ -0,0 +1,590 @@
+*/
+
+import cpp
+
+private import semmle.code.cpp.ir.IR
+private import semmle.code.cpp.controlflow.IRGuards
+private import semmle.code.cpp.ir.ValueNumbering
+private import RangeUtils
+private import SignAnalysis
+import Bound
+
+cached private module RangeAnalysisCache {
+
+  cached module RangeAnalysisPublic {
+    /**
+     * Holds if `b + delta` is a valid bound for `i`.
+     * - `upper = true`  : `i <= b + delta`
+     * - `upper = false` : `i >= b + delta`
+     *
+     * The reason for the bound is given by `reason` and may be either a condition
+     * or `NoReason` if the bound was proven directly without the use of a bounding
+     * condition.
+     */
+    cached predicate boundedInstruction(Instruction i, Bound b, int delta, boolean upper, Reason reason) {
+      boundedInstruction(i, b, delta, upper, _, _, reason)
+    }
+    
+    /**
+     * Holds if `b + delta` is a valid bound for `op`.
+     * - `upper = true`  : `op <= b + delta`
+     * - `upper = false` : `op >= b + delta`
+     *
+     * The reason for the bound is given by `reason` and may be either a condition
+     * or `NoReason` if the bound was proven directly without the use of a bounding
+     * condition.
+     */
+    cached predicate boundedOperand(Operand op, Bound b, int delta, boolean upper, Reason reason) {
+      boundedNonPhiOperand(op, b, delta, upper, _, _, reason)
+      or
+      boundedPhiOperand(op, b, delta, upper, _, _, reason)
+    }
+  }
+
+  /**
+   * Holds if `guard = boundFlowCond(_, _, _, _, _) or guard = eqFlowCond(_, _, _, _, _)`.
+   */
+  cached predicate possibleReason(IRGuardCondition guard) {
+    guard = boundFlowCond(_, _, _, _, _)
+    or
+    guard = eqFlowCond(_, _, _, _, _)
+  }
+
+}
+private import RangeAnalysisCache
+import RangeAnalysisPublic
+
+/**
+ * Gets a condition that tests whether `i` equals `bound + delta` at `use`.
+ *
+ * If the condition evaluates to `testIsTrue`:
+ * - `isEq = true`  : `i == bound + delta`
+ * - `isEq = false` : `i != bound + delta`
+ */
+private IRGuardCondition eqFlowCond(Operand op, Operand bound, int delta,
+  boolean isEq, boolean testIsTrue)
+{
+  exists(Operand compared |
+    result.ensuresEq(compared, bound, delta, op.getInstruction().getBlock(), isEq) and
+    result.controls(bound.getInstruction().getBlock(), testIsTrue) and
+    valueNumber(compared.getDefinitionInstruction()) = valueNumber (op.getDefinitionInstruction())
+  )
+}
+
+/**
+ * Holds if `op1 + delta` is a valid bound for `op2`.
+ * - `upper = true`  : `op2 <= op1 + delta`
+ * - `upper = false` : `op2 >= op1 + delta`
+ */
+private predicate boundFlowStepSsa(
+  NonPhiOperand op2, Operand op1, int delta, boolean upper, Reason reason
+) {
+  op2.getDefinitionInstruction().getAnOperand().(CopySourceOperand) = op1 and
+  (upper = true or upper = false) and
+  reason = TNoReason() and
+  delta = 0
+  or
+  exists(IRGuardCondition guard |
+    guard = boundFlowCond(op2, op1, delta, upper, _) and
```

Comment:
It looks like the conjunct `result.controls(op.getInstruction().getBlock(), testIsTrue)` has generally been folded into `boundFlowCond` when comparing to the java version.  Was there a particular reason for this?

Reply:
That's because my `boundFlowCond` takes an `Operand` rather than an `Instruction`; I think it doesn't make sense to talk about an `Operand` being bounded by a guard that doesn't control it. It's possible that I should change `boundFlowCond` to take an `Instruction` and check the guard in `boundFlowStepSsa`.

Reply:
Right, but then the `boundFlowCond` predicate is actually part of the family of edge-defining predicates, so it should be named something including `step`, and the last parameter is no longer needed.  In fact it becomes more or less equal to `boundFlowStepSsa`, so it could be inlined.  Maybe collect these two predicates into one named something like `boundFlowStepCond` so it's clear that it's responsible for defining the edges that arise from conditions.

Reply:
I'm going with the second suggestion; it simplifies eqFlowCond as well